core/src/main/scala/org/apache/spark/sql/execution/joins/HashJoinExec.scala

```diff
@@ -183,17 +181,45 @@ case class HashJoinExec(leftKeys: Seq[Expression],
   // return empty here as code of required variables is explicitly instantiated
   override def usedInputs: AttributeSet = AttributeSet.empty
 
+  private def findShuffleDeps(buildRDDs: Seq[RDD[_]]) : Seq[Dependency[_]] = {
+    val buildShuffleDeps: Seq[Dependency[_]] = buildRDDs.flatMap(_.dependencies
+      .filter(_.isInstanceOf[ShuffleDependency[_, _, _]]))
+    val oneToOneRdds = buildRDDs.flatMap(_.dependencies).
+      filter(_.isInstanceOf[OneToOneDependency[_]]).map(_.rdd)
+    if (oneToOneRdds.isEmpty) {
+      buildShuffleDeps
+    } else {
+      buildShuffleDeps ++ findShuffleDeps(oneToOneRdds)
+    }
+  }
```

Comment:
If you check the added test, there are two problems that it exposes: 

1. If the bottom most node in the build side is a replicated table, the join is considered a replicated table join. So if the build side is a replicated table with a shuffle, the shuffled rdd needs to be materialized. We were not doing that previously. 

2. If the buildPlan has Filter as the top most node, the buildRDD becomes a MapPartitionsRDD instead of ShuffledRDD and hence the following check returns no dependencies.

buildRDDs.flatMap(_.dependencies.filter(_.isInstanceOf[ShuffleDependency[_, _, _]]))

So, during execution, the build side reutrns 0 results because the inner ShuffledRDD is not materialized. However, in the test, the ShuffledRDD is wrapped inside the MapPartitionsRDD. We need to go inside the MapPartitionsRDD to find out the dependencies to materialize the internal ShuffledRDD. 

I have merged the latest master on this branch, in case you want to try it out.